src/Network/NWProtocolStack.cs

```diff
@@ -51,9 +51,28 @@ static void TrampolineIterateHandler (IntPtr block, IntPtr options)
 		{
 			var del = BlockLiteral.GetTarget<Action<NWProtocolOptions>> (block);
 			if (del != null) {
-				var x = new NWProtocolOptions (options, owns: false);
-				del (x);
-				x.Dispose ();
+				NWProtocolOptions castedOptions;
+				using (var tempOptions = new NWProtocolOptions (options, owns: false)) 
+				using (var definition = tempOptions.ProtocolDefinition) {
+					if (definition.Equals (NWProtocolDefinition.TcpDefinition)) {
+						castedOptions = new NWProtocolTcpOptions (options, owns: false);
+					}
+					if (definition.Equals (NWProtocolDefinition.UdpDefinition)) {
+						castedOptions = new NWProtocolUdpOptions (options, owns: false);
+					} 
+					if (definition.Equals (NWProtocolDefinition.TlsDefinition)) {
+						castedOptions = new NWProtocolTlsOptions (options, owns: false);
+					}
+					if (definition.Equals (NWProtocolDefinition.IPDefinition)) {
+						castedOptions = new NWProtocolIPOptions (options, owns: false);
+					}
+					if (definition.Equals (NWProtocolDefinition.WebSocketDefinition)) {
+						castedOptions = new NWWebSocketOptions (options, owns: false);
+					}
+					castedOptions = null;
+				}
+				del (castedOptions);
```

Comment:
should you call `del` with `null` ?!?

Reply:
I would to notify that we found something, although maybe returning the abstract class is a much better idea. More useful than null.